File: 1315-sum-of-nodes-with-even-valued-grandparent/1315-sum-of-nodes-with-even-valued-grandparent.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static class TreeNodeWithParent {
        TreeNodeWithParent parent;
        int val;
        TreeNode left;
        TreeNode right;
        TreeNodeWithParent(TreeNode node, TreeNodeWithParent parent){
            this.val = node.val;
            this.left = node.left;
            this.right = node.right;
            this.parent = parent;
        }
        TreeNodeWithParent(TreeNode node){
            this.val = node.val;
            this.left = node.left;
            this.right = node.right;
            parent = null;
        }
    }
    public int sumEvenGrandparent(TreeNode root) {
        if(root == null){
            return 0;
        }
        int total = 0;
        Queue<TreeNodeWithParent> queue = new LinkedList<>();
        queue.add(new TreeNodeWithParent(root));
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i=0;i<size;i++){
                TreeNodeWithParent node = queue.poll();
                if(node.parent!=null && node.parent.parent!=null && (node.parent.parent.val&1)==0){
                    total+=node.val;
                }
                if(node.left!=null){
                    queue.add(new TreeNodeWithParent(node.left, node));
                }
                if(node.right!=null){
                    queue.add(new TreeNodeWithParent(node.right, node));
                }
            }
        }
        return total;
                //    6,
                //   7,8,
                // 2,7,1,3,
        //9,null,1,4,null,null,null,5
    }
}

File: 1315-sum-of-nodes-with-even-valued-grandparent/1315-sum-of-nodes-with-even-valued-grandparent.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sumEvenGrandparent(root: TreeNode?): Int {
        if(root == null){
            return 0
        }
        var total = 0
        val queue: Queue<TreeNode?> = LinkedList()
        queue.add(root)
        while(!queue.isEmpty()){
            
            val size = queue.size
            for(i in 0 until size){
                val node = queue.poll()
                if(node == null){
                    continue
                }
                val isNodeEven = (node.`val` and 1)==0
                if(node.left!=null){
                    queue.add(node.left)
                    if(isNodeEven){
                        if(node?.left?.left != null){
                            total += node.left.left.`val`
                        }
                        if(node?.left?.right != null){
                            total += node.left.right.`val`
                        }
                    }
                }
                if(node.right!=null){
                    queue.add(node.right)
                    if(isNodeEven){
                        if(node?.right?.left != null){
                            total += node.right.left.`val`
                        }
                        if(node?.right?.right != null){
                            total += node.right.right.`val`
                        }
                    }
                }
            }
        }
        return total
    }
}